app/models/manageiq/providers/redhat/infra_manager/vm.rb

```diff
@@ -102,4 +107,103 @@ def has_required_host?
   def self.calculate_power_state(raw_power_state)
     POWER_STATES[raw_power_state] || super
   end
+
+  #
+  # This method is executed in the UI worker.
+  # It adds a task to calculate the URL of oVirt UI for this virtual machine.
+  #
+  # @return [Integer] The identifier of the task.
+  #
+  def queue_generate_ui_url
+    # Create the task options:
+    task_options = {
+      action: "Generate oVirt UI URL for VM '#{name}'"
+    }
+
+    # Create the queue options:
+    queue_options = {
+      class_name: self.class.name,
+      instance_id: id,
+      method_name: 'generate_ui_url',
+      priority: MiqQueue::HIGH_PRIORITY,
+      role: 'ems_operations',
+      zone: my_zone,
+      args: []
+    }
+
+    # Add the task to the queue and return the identifier:
+    MiqTask.generic_action_with_callback(task_options, queue_options)
+  end
+
+  #
+  # This method is executed in the provider worker.
+  # It calculates the URL of oVirt UI using the credentials of the provider.
+  #
+  # @return [String] The oVirt URL.
+  #
+  def generate_ui_url
+    # Get the reference to the EMS:
+    ems = ext_management_system
+
+    # Get the credentials of the provider:
+    user = ems.authentication_userid(:default)
+    password = ems.authentication_password(:default)
+
+    # Get the oVirt Engine endpoint:
+    endpoint = ems.default_endpoint
+
+    # TODO(vs) use standard Ruby Net::HTTP
+    #   https://apidock.com/ruby/Net/HTTP
+
+    # Create the HTTP client:
+    insecure = endpoint.verify_ssl == OpenSSL::SSL::VERIFY_NONE,
+    ca_certs = endpoint.certificate_authority
+    ca_certs = nil if ca_certs.blank?
+    ca_certs = [ca_certs] if ca_certs
+    client = OvirtSDK4::HttpClient.new(
+      insecure: insecure,
+      ca_certs: ca_certs,
+      cookies: true,
+      log: $rhevm_log
+    )
+
+    # Request new SSO token using provider's credentials:
+    ssoTokenScope = 'ovirt-app-admin%20ovirt-app-portal'
+    ssoTokenUrl = URI::HTTPS.build(
+      host: endpoint.hostname,
+      port: endpoint.port,
+      path: '/ovirt-engine/sso/oauth/token',
+      query: "grant_type=password&username=#{user}&password=#{password}&scope=#{ssoTokenScope}"
+    ).to_s
+    request = OvirtSDK4::HttpRequest.new
+    request.method = :GET
+    request.url = ssoTokenUrl
+    request.headers = { 'Accept': 'application/json' }
```

Comment:
Would be nice to have a `User-Agent` header, so that looking at the web server logs of the oVirt engine it is easy to identify this as a request coming from the ManageIQ oVirt provider.